bindings/python/crocoddyl/core/state-base.hpp

```diff
@@ -108,54 +103,93 @@ class StateAbstract_wrap : public StateAbstract, public bp::wrapper<StateAbstrac
                    << "x1 has wrong dimension (it should be " + std::to_string(nx_) + ")");
     }
 
-    if (firstsecond == "both") {
-      bp::list Jacs =
-          bp::call<bp::list>(this->get_override("Jdiff").ptr(), (Eigen::VectorXd)x0, (Eigen::VectorXd)x1, firstsecond);
-      return Jacs;
-    } else {
-      Eigen::MatrixXd J = bp::call<Eigen::MatrixXd>(this->get_override("Jdiff").ptr(), (Eigen::VectorXd)x0,
-                                                    (Eigen::VectorXd)x1, firstsecond);
-      bp::list list;
-      list.append(J);
-      return list;
-    }
-  }
-
-  void Jintegrate(const Eigen::Ref<const Eigen::VectorXd>& x, const Eigen::Ref<const Eigen::VectorXd>& dx,
-                  Eigen::Ref<Eigen::MatrixXd> Jfirst, Eigen::Ref<Eigen::MatrixXd> Jsecond,
-                  Jcomponent _firstsecond) const {
-    std::string firstsecond;
-    switch (_firstsecond) {
+    bp::list Jacs;
+    switch (firstsecond) {
       case first: {
-        firstsecond = "first";
+        Eigen::MatrixXd J = bp::call<Eigen::MatrixXd>(this->get_override("Jdiff").ptr(), (Eigen::VectorXd)x0,
+                                                      (Eigen::VectorXd)x1, firstsecond);
+        Jacs.append(J);
         break;
       }
       case second: {
-        firstsecond = "second";
+        Eigen::MatrixXd J = bp::call<Eigen::MatrixXd>(this->get_override("Jdiff").ptr(), (Eigen::VectorXd)x0,
+                                                      (Eigen::VectorXd)x1, firstsecond);
+        Jacs.append(J);
         break;
       }
       case both: {
-        firstsecond = "both";
+        Jacs = bp::call<bp::list>(this->get_override("Jdiff").ptr(), (Eigen::VectorXd)x0, (Eigen::VectorXd)x1,
+                                  firstsecond);
+        break;
+      }
+      default: {
+        Jacs = bp::call<bp::list>(this->get_override("Jdiff").ptr(), (Eigen::VectorXd)x0, (Eigen::VectorXd)x1,
+                                  firstsecond);
         break;
       }
-      default: { firstsecond = "both"; }
     }
+    return Jacs;
+  }
 
+  void Jintegrate(const Eigen::Ref<const Eigen::VectorXd>& x, const Eigen::Ref<const Eigen::VectorXd>& dx,
+                  Eigen::Ref<Eigen::MatrixXd> Jfirst, Eigen::Ref<Eigen::MatrixXd> Jsecond,
+                  const Jcomponent firstsecond, const AssignmentOp op) const {
     bp::list res = Jintegrate_wrap(x, dx, firstsecond);
```

Comment:
Why don't you call Jintegrate_wrap with the op argument, instead of doing the + and - operations again _inside_ the bindings?

Reply:
The Python user defines the Jacobian computations, then the operators are defined in `StateAbstract_wrap` class. This makes possible to call Python code and use any operator from the c++ interface. However, Python derived classes don't have the concept of operators.

Unfortunately, it is not possible to bind virtual functions with Eigen::Ref types (at least yet). Note that in Pyhton `Jintegrate` doesn't have as arguments `Jfirst` and `Jsecond`, and Pinocchio follows the same approach.

We could change this in future (if we believe it's nicer), but this will require some extra work from Eigenpy side (@jcarpent -- you might want to see this comment).

On other hand, I couldn't add a python test for `JIntegrateTransport` since we need the bindings in Pinocchio.

Reply:
@cmastalli Again, there is no reason that EigenPy cannot do what you mention. It would be useful to provide some code examples to support your issue. Otherwise, it is totally unclear for the readers.

Reply:
Let me try it again. First of all, @proyan suggestion requires to change the Python API, i.e.
```python
import crocoddyl
s = crocoddyl.StateVector(3)
x, dx = s.rand(), s.rand()
# Current API
s.Jintegrate(x, dx, crocoddyl.Jcomponent.both) # we don't pass Jfirst / Jsecond
# New API required to include operators
Jfirst = np.array((s.ndx, s.ndx))
s.Jintegrate(x, dx, crocoddyl.Jcomponent.both, Jfirst, Jsecond) # note this code is not availble
```

We could change the API, but let's have a time to discuss it. I would avoid this change here, I am not sure if we really need to support operators in Python. Again, the operators are supported in c++ interface even using Python derived code.

**Regarding Eigen::Ref**
For binding a virtual function I need to type cast the `Eigen::Ref` thing, e.g.
```cpp
bp::call<void>(this->get_override("Jintegrate").ptr(), (Eigen::VectorXd)x,
                                                      (Eigen::VectorXd)dx, firstsecond);
```
because I am not able to call this function from a Python Dict of Numpy objects; it doesn't found a value converter.

However, I agree with you that it is possible to solve this problem, and I need time to figure out how to do it.

Reply:
@cmastalli Please provide a minimal and repoducible example, isolated from Croco.

Reply:
@jcarpent I will work later on it.
Let's keep this for further iterations. I don't want to block more this PR.